coderd/agentapi/metadata_test.go

```diff
@@ -0,0 +1,275 @@
+package agentapi_test
+
+import (
+	"context"
+	"encoding/json"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"cdr.dev/slog/sloggers/slogtest"
+
+	agentproto "github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/coderd/agentapi"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbmock"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/coderd/database/pubsub"
+)
+
+func TestBatchUpdateMetadata(t *testing.T) {
+	t.Parallel()
+
+	agent := database.WorkspaceAgent{
+		ID: uuid.New(),
+	}
+
+	t.Run("OK", func(t *testing.T) {
+		t.Parallel()
+
+		dbM := dbmock.NewMockStore(gomock.NewController(t))
+		pub := pubsub.NewInMemory()
+
+		now := dbtime.Now()
+		req := &agentproto.BatchUpdateMetadataRequest{
+			Metadata: []*agentproto.Metadata{
+				{
+					Key: "awesome key",
+					Result: &agentproto.WorkspaceAgentMetadata_Result{
+						CollectedAt: timestamppb.New(now.Add(-10 * time.Second)),
+						Age:         10,
+						Value:       "awesome value",
+						Error:       "",
+					},
+				},
+				{
+					Key: "uncool key",
+					Result: &agentproto.WorkspaceAgentMetadata_Result{
+						CollectedAt: timestamppb.New(now.Add(-3 * time.Second)),
+						Age:         3,
+						Value:       "",
+						Error:       "uncool value",
+					},
+				},
+			},
+		}
+
+		dbM.EXPECT().UpdateWorkspaceAgentMetadata(gomock.Any(), database.UpdateWorkspaceAgentMetadataParams{
+			WorkspaceAgentID: agent.ID,
+			Key:              []string{req.Metadata[0].Key, req.Metadata[1].Key},
+			Value:            []string{req.Metadata[0].Result.Value, req.Metadata[1].Result.Value},
+			Error:            []string{req.Metadata[0].Result.Error, req.Metadata[1].Result.Error},
+			// The value from the agent is ignored.
+			CollectedAt: []time.Time{now, now},
+		}).Return(nil)
+
+		api := &agentapi.MetadataAPI{
+			AgentFn: func(context.Context) (database.WorkspaceAgent, error) {
+				return agent, nil
+			},
+			Database: dbM,
+			Pubsub:   pub,
+			Log:      slogtest.Make(t, nil),
+			TimeNowFn: func() time.Time {
+				return now
+			},
+		}
+
+		// Watch the pubsub for events.
+		var (
+			eventCount int64
+			gotEvent   agentapi.WorkspaceAgentMetadataChannelPayload
+		)
+		cancel, err := pub.Subscribe(agentapi.WatchWorkspaceAgentMetadataChannel(agent.ID), func(ctx context.Context, message []byte) {
+			if atomic.AddInt64(&eventCount, 1) > 1 {
+				return
+			}
+			require.NoError(t, json.Unmarshal(message, &gotEvent))
+		})
+		require.NoError(t, err)
+		defer cancel()
+
+		resp, err := api.BatchUpdateMetadata(context.Background(), req)
+		require.NoError(t, err)
+		require.Equal(t, &agentproto.BatchUpdateMetadataResponse{}, resp)
+
+		require.Equal(t, int64(1), atomic.LoadInt64(&eventCount))
```

Comment:
Pubsub calls the subscriber on a goroutine, which means that this test can race.

You could create a fake pubsub that only implements `Publish()` (the rest can panic), and stores the results internally, then assert on the state of the fake pubsub.

An even better change would be to break up the Pubsub interface like

```
type Publisher interface {
    Publish(event string, message []byte) error
}

type Subscriber interface {
   Subscribe(event string, listener Listener) (cancel func(), err error)
   SubscribeWithErr(event string, listener ListenerWithErr) (cancel func(), err error)
}

type Pubsub interface {
    Publisher
    Subscriber
    io.Closer
}
```

then the `MetadataAPI` can take a `Publisher` and you only need to implement `Publish()` on your fake.